File: js/호텔_방_배정.js
function solution(k, room_number) {
    const answer = [];
    const rooms = new Map();
    
    for (let want of room_number) {
        answer.push(findRoom(want));
    }
    
    return answer;

    function findRoom(want) {
        if (!rooms.has(want)) {
            rooms.set(want, want + 1);
            return want;
        }
        const next = findRoom(rooms.get(want));
        rooms.set(want, next + 1);
        return next;
    }
}


const k = 10;
const room_number = [1, 3, 4, 1, 3, 1];

solution(k, room_number)

File: python/kakao대비/호텔_방_배정.py
import sys
sys.setrecursionlimit(10 ** 6)

def solution(k, room_number):
    room_info = {}

    def reserve_room(n):
        if n in room_info:
            _next = reserve_room(room_info[n])
            room_info[n] = _next + 1
            return _next
        else:
            room_info[n] = n+1
            return n 
    
    return [reserve_room(n) for n in room_number]